include/sys/icu.h

```diff
@@ -0,0 +1,36 @@
+#ifndef _ICU_METHODS_H_
+#define _ICU_METHODS_H_
+
+#include <sys/device.h>
+#include <sys/resource.h>
+#include <sys/interrupt.h>
+
+typedef struct icu_methods {
+  void (*intr_setup)(device_t *dev, resource_t *irq, ih_filter_t *filter,
+                     ih_service_t *service, void *arg, const char *name);
+  void (*intr_teardown)(device_t *dev, resource_t *irq);
+} icu_methods_t;
+
+#define ICU_METHODS(dev) (*(icu_methods_t *)(dev)->driver->interfaces[DIF_ICU])
+
+static inline void icu_intr_setup(device_t *dev, resource_t *irq,
+                                  ih_filter_t *filter, ih_service_t *service,
+                                  void *arg, const char *name) {
+  assert(dev->parent);
+  if (!dev->parent->driver->interfaces[DIF_ICU]) {
```

Comment:
Maybe add something like 
```C
#define DIF_IMPLEMENST(dev, interface) ((dev)->interfaces[(interface)] != NULL)
```

Reply:
It's not a secret interfaces are stored simply as fields of an array, as simple array initialization is the way they are being assigned to drivers. I think that such macro would only obfuscate the code. Checking for null pointer is a very common pattern in C.

I believe the most elegant solution would be to change `..._METHODS` macros so they wouldn't dereference structs. Then, this check could be converted to something like
```
if(!ICU_METHODS(dev->parent))
```

Tbh I never liked that dereferencing inside those macros. It just seemed weird as most of the dereferencing in the code base seems to be done explicitly.

Reply:
I agree. We should change `..._METHODS` definitions. I also think that the following control flow is more coherent
```C
icu_intr_...(dev, ...) {
  device_t *parent = dev->parent;
  assert(parent);
  if (!ICU_METHODS(parent))
    icu_intr_...(parent, ...);
  else
    ICU_METHODS(parent)->intr_...(dev, ...);
}
```
This removes the inner `return` instruction.

Reply:
I changed `..._METHODS` macros which made the check cleaner.